cedar-policy-validator/src/entity_manifest.rs

```diff
@@ -0,0 +1,1306 @@
+    children: Fields<T>,
+    /// For entity types, this boolean may be `true`
+    /// to signal that all the ancestors in the entity hierarchy
+    /// are required (transitively).
+    ancestors_required: bool,
+    /// Optional data annotation, usually used for type information.
+    #[serde(skip_serializing, skip_deserializing)]
+    #[serde(bound(deserialize = "T: Default"))]
+    data: T,
+}
+
+/// A data path that may end with requesting the parents of
+/// an entity.
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct AccessPath {
+    /// The root variable that begins the data path
+    pub root: EntityRoot,
+    /// The path of fields of entities or structs
+    pub path: Vec<SmolStr>,
+    /// Request all the parents in the entity hierarchy of this entity.
+    pub ancestors_required: bool,
+}
+
+/// Entity manifest computation does not handle the full
+/// cedar language. In particular, the policies must follow the
+/// following grammar:
+/// ```text
+/// <expr> = <datapath-expr>
+///          <datapath-expr> in <expr>
+///          <expr> + <expr>
+///          if <expr> { <expr> } { <expr> }
+///          ... all other cedar operators not mentioned by datapath-expr
+
+/// <datapath-expr> = <datapath-expr>.<field>
+///                   <datapath-expr> has <field>
+///                   <variable>
+///                   <entity literal>
+/// ```
+/// The `get_expr_path` function handles `datapath-expr` expressions.
+/// This error message tells the user not to use certain operators
+/// before accessing record or entity attributes, breaking this grammar.
+// CAUTION: this type is publicly exported in `cedar-policy`.
+// Don't make fields `pub`, don't make breaking changes, and use caution
+// when adding public methods.
+#[derive(Debug, Clone, Error, Hash, Eq, PartialEq)]
+#[error("for policy `{policy_id}`, failed to analyze expression while computing entity manifest`")]
+pub struct FailedAnalysisError {
+    /// Source location
+    source_loc: Option<Loc>,
+    /// Policy ID where the error occurred
+    policy_id: PolicyID,
+    /// The kind of the expression that was unexpected
+    expr_kind: ExprKind<Option<Type>>,
+}
+
+impl Diagnostic for FailedAnalysisError {
+    impl_diagnostic_from_source_loc_opt_field!(source_loc);
+
+    fn help<'a>(&'a self) -> Option<Box<dyn Display + 'a>> {
+        Some(Box::new(format!(
+            "failed to compute entity manifest: {} operators are not allowed before accessing record or entity attributes",
+            self.expr_kind.operator_description()
+        )))
+    }
+}
+
+/// Error when expressions are partial during entity
+/// manifest computation
+// CAUTION: this type is publicly exported in `cedar-policy`.
+// Don't make fields `pub`, don't make breaking changes, and use caution
+// when adding public methods.
+#[derive(Debug, Clone, Error, Hash, Eq, PartialEq)]
+#[error(
+    "Entity slicing requires fully concrete policies. Got a policy with an unknown expression."
```

Comment:
Our convention is that all error messages start with a lowercase letter, and do not end with punctuation
```suggestion
    "entity slicing requires fully concrete policies. Got a policy with an unknown expression"
```